Substantially improved fb2k startup and script initialization

kbuffington opened the pull request:
...by deferring call to InitializeSubsytems to the on_init() callback.

See full discussion [here](https://hydrogenaud.io/index.php?topic=114686.175) and discussion of fix [here](https://hydrogenaud.io/index.php?topic=114686.msg997642#msg997642).

This _seems_ safe to do, but maybe there's a reason it must be done in DLL_PROCESS_ATTACH that I'm unaware of.

marc2k3 commented:
Included samples such as mine and Brett's JSplaylist bork on startup in default UI.

```
foo_spider_monkey_panel:
Error: Spider Monkey Panel v1.4.2-dev+ce39bb09 (Last.fm Bio: Last.fm Bio by marc2003)
include failed:
Font failed:
Failed to create GdiPlus object

File: helpers.js
Line: 644, Column: 23
Stack trace:
  @helpers.js:644:23
  @<main>:5:1

foo_spider_monkey_panel:
Error: Spider Monkey Panel v1.4.2-dev+ce39bb09 (JSPlaylist: JSPlaylist vSMP-Mod by Br3tt)
include failed:
CreateImage failed:
Failed to create GdiPlus object

File: WSHinputbox.js
Line: 6, Column: 15
Stack trace:
  @WSHinputbox.js:6:15
  @<main>:12:2
  @<main>:11:12
```

Reloading the panel works so my guess is Gdiplus startup needs to be done earlier.

Incidentally, you can avoid saving HINSTANCE from dllmain - just use core_api::get_my_instance().

TheQwertiest commented:
There is a reason why those are in DLL_PROCESS_ATTACH and not in `on_init`: `on_init` is called much later than component has actually started working. Hence some parts of component might not work at all with this change (see errors that @marc2k3 mentioned).

TheQwertiest commented:
Case in point: `core_api::get_my_instance` will return `null` if called too early.

TheQwertiest commented:
A proper deferring would be to call all the required init methods when they are actually needed through some sort of wrapper, but that will require much more work.

marc2k3 commented:
FWIW, I use initquit for setting up/nuking WTL and Scintilla. I've been doing this for awhile for other reasons - I have direct write enabled in my Scintilla control and this generates crash logs if you release scintilla resources from DLL_PROCESS_DETACH like the old code did for years.

https://github.com/marc2k3/foo_jscript_panel/blob/bfb1245e6d45fe06f2c9e69a064d2d462df4351c/src/foo_jscript_panel.cpp#L25

I also init Gdiplus even earlier outside of dllmain using this helper from libPPUI.

https://github.com/marc2k3/foo_jscript_panel/blob/bd0b7ad49e843f69e74bde5cfb975cc1814a83ad/src/foo_jscript_panel.cpp#L12
https://github.com/marc2k3/foobar2000-sdk/blob/7777c6afdd2cf1de708a65ab3ebd130a482d0db4/libPPUI/gdiplus_helpers.h#L16

kbuffington commented:
Welp, it seemed too good to be true.

_Knew_ I should have checked to see what jscript_panel was doing on startup, but I was too lazy to clone first 😆

I tried this on a bunch of scripts last night and everything worked fine but was doing all testing under CUI. Tried this morning in CUI and DUI and can't reproduce crashes using JSPlaylist or Last.fm bio. Wondering if maybe my slow machine (which causes the original problem) is masking this by allowing the DLLs to fully load first?

I'll take another crack at it. Since doesn't seem like @TheQwertiest is seeing this issue I'm probably in best position to see what fixes it.

marc2k3 commented:
Only tested with JSP but it seems initalising gdiplus is not required at all for CUI panels - only DUI.

kbuffington commented:
So CUI must already be initializing gdiplus I guess. Still weird that I can't reproduce in DUI, but again probably a race condition. Do the crashes happen every time for you?

marc2k3 commented:
I nuked the component so can't check now. 

Other info, my foobar startup is always rapid because I have a tiny collection and fb2k and all music is on SSD.

kbuffington commented:
I removed all components except for FSM, removed my entire music library and only left the default playlist. Still can't get FSM to crash on JSPlaylist. 

On the bright side, foobar starts in under a second!

kbuffington commented:
Okay new version. Marc says this one is working no crashes. I kept half of the startup functionality in the original InitializeSubsytems and moved Scintilla, WTL, ATL, and RichText to InitializeDelayedSubsystems. Thinking is that Scintilla, + the three GUI related libraries don't need to be loaded until there's actually a UI to display them in. It feels safe and I _believe_ I'm still seeing speed improvements. Gonna ask for more testing from some other people explaining the same issues.

marc2k3 commented:
Got bored so I downloaded the SMP source and setup the VS C++ profiler. There really isn't much going on during dllmain with the latest build from master.

My startup time as reported by console...

```
Components loaded in: 0:00.015567
Configuration read in: 0:00.004448
foobar2000 v1.6.6 beta 7 [portable]
Watching: E:\Music
Spider Monkey Panel v1.4.2-dev+ce39bb09 (JS Smooth Browser: JS Smooth Browser by Br3tt aka Falstaff): initialized in 83 ms
Spider Monkey Panel v1.4.2-dev+ce39bb09 (Library Tree: Library Tree v2.1.3 by WilB): initialized in 36 ms
Spider Monkey Panel v1.4.2-dev+ce39bb09 (JSPlaylist: JSPlaylist vSMP-Mod by Br3tt): initialized in 173 ms
Spider Monkey Panel v1.4.2-dev+ce39bb09 (Track Info + Seekbar + Buttons: Track Info + Seekbar + Buttons by marc2003): initialized in 49 ms
User Interface initialized in: 0:00.492207
FFmpeg version: 4.3.1
Startup time : 0:00.529246
```

![image](https://user-images.githubusercontent.com/62050527/118382438-7b926380-b5ed-11eb-969b-b780d4189975.png)

kbuffington commented:
I've been gone all day so wasn't able to post this until now, but I realized really late last night that there was no benefit from this version of the component, and _probably_ none from the first one I posted. I'm about 95% sure the slowdown is related to Media Library initialization. I'm gonna post a thread about it at some point because I'm curious but in some situations on startup FB2k just seems to know that nothing has changed and instantly shows "monitoring" as the status and other times it takes a while to read the library. Whenever it leads to reading the library it causes a slowdown. If your script is fast enough it might finish execution before, and if not, well you wait 20-30 seconds before the UI appears.
